lib/src/web_ui/code_input.dart

```diff
@@ -0,0 +1,164 @@
+library web_ui.code_input;
+
+import 'dart:async';
+import 'dart:html';
+
+import 'package:cs61a_scheme/cs61a_scheme.dart';
+
+import 'highlight.dart';
+
+const List<SchemeSymbol> noIndentForms = [
+  SchemeSymbol("let"),
+  SchemeSymbol("define"),
+  SchemeSymbol("lambda"),
+  SchemeSymbol("define-macro")
+];
+
+class CodeInput {
+  Element element;
+  bool _active = true;
+  final List<StreamSubscription> _subs = [];
+
+  Function(String code) runCode;
+  Function(int parens) parenListener;
+
+  CodeInput(Element log, this.runCode, {this.parenListener}) {
+    element = SpanElement()
+      ..classes = ['code-input']
+      ..contentEditable = 'true';
+    _subs.add(element.onKeyPress.listen(_onInputKeyPress));
+    _subs.add(element.onKeyDown.listen(_keyListener));
+    _subs.add(element.onKeyUp.listen(_keyListener));
+    log.append(element);
+    element.focus();
+    parenListener ??= (_) => null;
+    parenListener(missingParens);
+  }
+
+  String get text => element.text;
+
+  set text(String newText) {
+    element.text = newText;
+    highlight(atEnd: true);
+  }
+
+  bool get active => _active;
+
+  int get missingParens => countParens(text);
+
+  void deactivate() {
+    _active = false;
+    element.contentEditable = 'false';
+    for (var sub in _subs) {
+      sub.cancel();
+    }
+  }
+
+  Future highlight(
+      {bool saveCursor = false, int cursor, bool atEnd = false}) async {
+    if (saveCursor) {
+      await highlightSaveCursor(element);
+    } else if (cursor != null) {
+      await highlightCustomCursor(element, cursor);
+    } else if (atEnd) {
+      await highlightAtEnd(element, element.text);
+    } else {
+      element.innerHtml = highlightText(element.text);
+    }
+  }
+
+  Future _onInputKeyPress(KeyboardEvent event) async {
+    if ((missingParens ?? -1) > 0 &&
+        event.shiftKey &&
+        event.keyCode == KeyCode.ENTER) {
+      event.preventDefault();
+      element.text = element.text.trimRight() + ')' * missingParens + '\n';
+      runCode(element.text);
+      await delay(100);
+      await highlight();
+    } else if ((missingParens ?? -1) == 0 && event.keyCode == KeyCode.ENTER) {
+      event.preventDefault();
+      element.text = element.text.trimRight() + '\n';
+      runCode(element.text);
+      await delay(100);
+      await highlight();
+    } else if ((missingParens ?? 0) > 0 && KeyCode.ENTER == event.keyCode) {
+      event.preventDefault();
+      int cursor = _currPosition();
```

Comment:
I noticed that the `_currPosition` function is only used here. I wonder if it makes more sense to replace this with
```
int cursor = findPosition(element, window.getSelection().getRangeAt(0));
```

and remove the `_currPosition` function all together